imjoy/imjoyPluginEngine.py

```diff
@@ -731,6 +737,102 @@ def connect(sid, environ):
     logger.info("connect %s", sid)
 
 
+async def read_and_forward_terminal_output():
+    """ read from ternial and forward to the client
+    """
+    max_read_bytes = 1024 * 20
+    while True:
+        await asyncio.sleep(0.01)
+        if "fd" in terminal_session:
+            timeout_sec = 0
+            (data_ready, _, _) = select.select(
+                [terminal_session["fd"]], [], [], timeout_sec
+            )
+            if data_ready:
+                output = os.read(terminal_session["fd"], max_read_bytes).decode()
+                if len(output) > 0:
+                    await sio.emit("terminal_output", {"output": output})
+
+
+@sio.on("start_terminal", namespace=NAME_SPACE)
+async def on_start_terminal(sid, kwargs):
+    """new terminal client connected"""
+    try:
+        if sid not in registered_sessions:
+            logger.debug("client %s is not registered.", sid)
+            return {"success": False, "error": "client not registered."}
+
+        if "child_pid" in terminal_session:
+            # already started child process, don't start another
+            return {"success": True, "exists": True}
+        if sys.platform == "linux" or sys.platform == "linux2":
+            # linux
+            default_terminal_command = ["bash"]
+        elif sys.platform == "darwin":
+            # OS X
+            default_terminal_command = ["bash"]
+        elif sys.platform == "win32":
+            # Windows...
+            default_terminal_command = ["start", "cmd"]
+        else:
+            default_terminal_command = ["bash"]
+        cmd = kwargs.get("cmd", default_terminal_command)
+
+        # create child process attached to a pty we can read from and write to
+        (child_pid, fd) = pty.fork()
+        if child_pid == 0:
+            # this is the child process fork.
+            # anything printed here will show up in the pty, including the output
+            # of this subprocess
+            subprocess.run(cmd)
+        else:
+            # this is the parent process fork.
+            # store child fd and pid
+            terminal_session["fd"] = fd
+            terminal_session["child_pid"] = child_pid
+            set_winsize(fd, 50, 50)
+            cmd = " ".join(shlex.quote(c) for c in cmd)
+            print("child pid is", child_pid)
+            print(
+                f"starting background task with command `{cmd}` to continously read "
+                "and forward pty output to client"
+            )
+            logger.debug("xterm task started", terminal_session)
+            # os.write(terminal_session["fd"], "\r".encode())
+            asyncio.ensure_future(
+                read_and_forward_terminal_output(), loop=asyncio.get_event_loop()
```

Comment:
When is the task cancelled?

Reply:
It will start forever. Will fix when we support multiple terminal sessions.

Reply:
We don't need to clean up tasks when the plugin exits?

Reply:
This is not a plugin, it's an engine service, which shares the life cycle with the engine itself.

Reply:
I think we will have trouble when the terminal process accidentally closed. But let's leave it for now. We will handle it when we start to work on the multi-terminal session support.